libindy_vdr/src/pool/cache/storage.rs

```diff
@@ -0,0 +1,227 @@
+    fn first_key_value(&self) -> Option<(u128, V)> {
+        match self.first() {
+            Ok(Some((k, v))) => serde_json::from_slice(v.as_ref()).ok().map(|v| {
+                (
+                    u128::from_be_bytes(k.as_ref().try_into().unwrap_or([0; 16])),
+                    v,
+                )
+            }),
+            _ => None,
+        }
+    }
+    fn last_key_value(&self) -> Option<(u128, V)> {
+        match self.last() {
+            Ok(Some((k, v))) => serde_json::from_slice(v.as_ref()).ok().map(|v| {
+                (
+                    u128::from_be_bytes(k.as_ref().try_into().unwrap_or([0; 16])),
+                    v,
+                )
+            }),
+            _ => None,
+        }
+    }
+    fn get(&self, key: &u128) -> Option<V> {
+        match Tree::get(self, key.to_be_bytes()).map(|v| v) {
+            Ok(Some(v)) => serde_json::from_slice(v.as_ref()).ok(),
+            _ => None,
+        }
+    }
+    fn insert(&mut self, key: u128, value: V) -> Option<V> {
+        match Tree::insert(self, key.to_be_bytes(), serde_json::to_vec(&value).unwrap()) {
+            Ok(Some(v)) => serde_json::from_slice(v.as_ref()).ok(),
+            _ => None,
+        }
+    }
+    fn remove(&mut self, key: &u128) -> Option<V> {
+        match Tree::remove(self, key.to_be_bytes()).map(|v| v) {
+            Ok(Some(v)) => serde_json::from_slice(&v).ok(),
+            _ => None,
+        }
+    }
+    fn entries(&self) -> Box<dyn Iterator<Item = (u128, V)>> {
+        Box::new(self.iter().filter_map(|r| {
+            r.ok().and_then(|(k, v)| {
+                serde_json::from_slice(v.as_ref())
+                    .ok()
+                    .map(|v| (u128::from_be_bytes(k.as_ref().try_into().unwrap()), v))
+            })
+        }))
+    }
+}
+impl<O: Ord + Copy + Send + Sync, V: Clone + Send + Sync> OrderedStore<O, V> for BTreeMap<O, V> {
+    fn len(&self) -> usize {
+        BTreeMap::len(self)
+    }
+    fn first_key_value(&self) -> Option<(O, V)> {
+        BTreeMap::first_key_value(self).map(|(o, v)| (*o, v.clone()))
+    }
+    fn last_key_value(&self) -> Option<(O, V)> {
+        BTreeMap::last_key_value(self).map(|(o, v)| (*o, v.clone()))
+    }
+    fn get(&self, key: &O) -> Option<V> {
+        BTreeMap::get(self, key).map(|v| v.clone())
+    }
+    fn insert(&mut self, key: O, value: V) -> Option<V> {
+        BTreeMap::insert(self, key, value)
+    }
+    fn remove(&mut self, key: &O) -> Option<V> {
+        BTreeMap::remove(self, key)
+    }
+    fn entries(&self) -> Box<dyn Iterator<Item = (O, V)> + '_> {
+        Box::new(self.iter().map(|(o, v)| (o.clone(), v.clone())))
+    }
+}
+/// A hashmap that also maintains a BTreeMap of keys ordered by a given value
+/// This is useful for structures that need fast O(1) lookups, but also need to evict the oldest or least recently used entries
+/// The Ordered Store must contain both the keys and values for persistence
+pub struct OrderedHashMap<K, O, V>(
```

Comment:
I'm not sure why this is a tuple struct with a 2-tuple as its only member, instead of having two named fields?